pkg/erofs/erofs.go

```diff
@@ -0,0 +1,643 @@
+	}
+	if fiPre.Mode()&os.ModeSymlink != 0 {
+		return cmd, errors.Errorf("Refusing to mount onto a symbolic linkd")
+	}
+
+	// It would be nice to only enable debug (or maybe to only log to file at all)
+	// if 'stacker --debug', but we do not have access to that info here.
+	// to debug erofsfuse, use "allow_other,debug"
+	optionArgs := "allow_other,debug"
+	if notifyOpts != "" {
+		optionArgs += "," + notifyOpts
+	}
+	cmd = exec.Command(erofsFuseInfo.Path, "-f", "-o", optionArgs, squashFile, extractDir)
+	cmd.Stdin = nil
+	cmd.Stdout = cmdOut
+	cmd.Stderr = cmdOut
+	cmdOut.Write([]byte(fmt.Sprintf("# %s\n", strings.Join(cmd.Args, " "))))
+	if err != nil {
+		return cmd, errors.Wrapf(err, "Failed writing to %s", logf)
+	}
+	log.Debugf("Extracting %s -> %s with %s [%s]", squashFile, extractDir, erofsFuseInfo.Path, logf)
+	err = cmd.Start()
+	if err != nil {
+		return cmd, err
+	}
+
+	// now poll/wait for one of 3 things to happen
+	// a. child process exits - if it did, then some error has occurred.
+	// b. the directory Entry is different than it was before the call
+	//    to erofsfuse.  We have to do this because we do not have another
+	//    way to know when the mount has been populated.
+	//    https://github.com/vasi/erofsfuse/issues/49
+	// c. a timeout (timeLimit) was hit
+	startTime := time.Now()
+	timeLimit := 30 * time.Second
+	alarmCh := make(chan struct{})
+	go func() {
+		_ = cmd.Wait()
+		close(alarmCh)
+	}()
+	if erofsFuseInfo.SupportsNotfiy {
+		notifyCh := make(chan byte)
+		log.Infof("%s supports notify pipe, watching %q", erofsFuseInfo.Path, notifyPath)
+		go func() {
+			f, err := os.Open(notifyPath)
+			if err != nil {
+				return
+			}
+			defer f.Close()
+			b1 := make([]byte, 1)
+			for {
+				n1, err := f.Read(b1)
+				if err != nil {
+					return
+				}
+				if err == nil && n1 >= 1 {
+					break
+				}
+			}
+			notifyCh <- b1[0]
+		}()
+
+		select {
+		case <-alarmCh:
+			cmd.Process.Kill()
+			return cmd, errors.Wrapf(err, "Gave up on erofsfuse mount of %s with %s after %s", squashFile, erofsFuseInfo.Path, timeLimit)
+		case ret := <-notifyCh:
+			if ret == 's' {
+				return cmd, nil
+			} else {
+				return cmd, errors.Errorf("erofsfuse returned an error, check %s", logf)
+			}
+		}
+	}
+	for count := 0; !common.FileChanged(fiPre, extractDir); count++ {
+		if cmd.ProcessState != nil {
+			// process exited, the Wait() call in the goroutine above
+			// caused ProcessState to be populated.
+			return cmd, errors.Errorf("erofsfuse mount of %s with %s exited unexpectedly with %d", squashFile, erofsFuseInfo.Path, cmd.ProcessState.ExitCode())
+		}
+		if time.Since(startTime) > timeLimit {
+			cmd.Process.Kill()
+			return cmd, errors.Wrapf(err, "Gave up on erofsfuse mount of %s with %s after %s", squashFile, erofsFuseInfo.Path, timeLimit)
+		}
+		if count%10 == 1 {
+			log.Debugf("%s is not yet mounted...(%s)", extractDir, time.Since(startTime))
+		}
+		time.Sleep(time.Duration(50 * time.Millisecond))
```

Comment:
Any reason to sleep instead of just polling? we're waiting on an external process anyhow we're not doing anything else at the moment.